### PR TITLE
MultiSig - Latest Test-cases for NFTs modules

### PR DESCRIPTION
**Pull request by githubckgoh1439**
Hi,  Mostafa 

I added new test-cases for NFTs.
Please review and comments.

TQ.

### CODE REVIEW

File: tests/multisigNFTs_test.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+		{"bank", false, false, "top-up Multisig Group-address1 - commit", "multisig-acc-nft-1", "800400000cin", 0, bankInfo{"multisig-acc-nft-1", "nft-grp-addr-1", "10000000000cin"}, "MEMO : top-up account", nil},
+		{"bank", false, false, "top-up Multisig Group-address2 - commit", "multisig-acc-nft-2", "800400000cin", 0, bankInfo{"multisig-acc-nft-2", "nft-grp-addr-2", "10000000000cin"}, "MEMO : top-up account", nil},
+		{"bank", false, false, "top-up Multisig Group-address3 - commit", "multisig-acc-nft-3", "800400000cin", 0, bankInfo{"multisig-acc-nft-3", "nft-grp-addr-3", "10000000000cin"}, "MEMO : top-up account", nil},
+		{"bank", false, false, "top-up Multisig Group-address4 - commit", "multisig-acc-nft-4", "800400000cin", 0, bankInfo{"multisig-acc-nft-4", "nft-grp-addr-4", "10000000000cin"}, "MEMO : top-up account", nil},
+
+		//2. module : kyc_test
+		{"kyc", false, false, "Doing kyc - nft-mostafa - commit", "nft-kyc-auth-1", "0cin", 0, kycInfo{"nft-kyc-auth-1", "nft-kyc-issuer-1", "nft-kyc-prov-1", "whitelist", "nft-mostafa", "nft-mostafa", "testKyc123452222", "0"}, "", nil},
+
+		//3. module : maintenance_test
+		//add nft-mostafa as nonfungible authorised address
```
Comment (b00f) [inline tests/multisigNFTs_test.go:84]: This case doesn't have conflict with other (maintenance_tests.go)?

Comment (b00f) [reply tests/multisigNFTs_test.go:84]: Maybe better we run maintenance test first then multisig. Any idea?

File: tests/multisigNFTs_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+		{"multiSig", true, true, "[case-2.0] Transfer MultiSig Owner - Error, due to without KYC - wait-15-seconds.																																					 ", "multisig-nokyc", "100000000cin", 0, MultisigInfo{"transfer-ownership", "multisig-acc-nft-1", "multisig-nokyc", 0, nil, "nft-grp-addr-5", 0, nil}, "MEMO : Transfer MultiSig Owner.", nil},
+		{"multiSig", false, false, "[case-2.0] Transfer MultiSig Owner - [from multisig-acc-nft-1 to multisig-acc-nft-4] Happy Path - commit. - wait-15-seconds.																		 ", "multisig-acc-nft-1", "100000000cin", 0, MultisigInfo{"transfer-ownership", "multisig-acc-nft-1", "multisig-acc-nft-4", 0, nil, "nft-grp-addr-3", 0, nil}, "MEMO : Transfer MultiSig Owner.", nil},
+		{"multiSig", true, true, "[case-2.0] Re-transfer MultiSig Owner - Error, due to Owner of group address invalid [MultiSig-account already been transfer to others]. - wait-15-seconds.", "multisig-acc-nft-1", "100000000cin", 0, MultisigInfo{"transfer-ownership", "multisig-acc-nft-1", "multisig-acc-nft-4", 0, nil, "nft-grp-addr-3", 0, nil}, "MEMO : Transfer MultiSig Owner.", nil},
+
+		// signer without through KYC
```
Comment (b00f) [inline tests/multisigNFTs_test.go:187]: I think this case is redundant. We have tested this in multisig_test.go

File: tests/multisigNFTs_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		{"multiSig", false, false, "[case-2.0] Create MultiSig Tx for NFTs [Accept-token-ownership] - [internalTx3x_acceptTokenOwnership_err1] - wait-15-seconds.", "multisig-acc-nft-4", "100000000cin", 0, MultisigInfo{"create-internal-tx", "", "", 0, nil, "nft-grp-addr-4", 1, internalTx3x_acceptTokenOwnership_err1}, "MEMO : xxxxx", nil}, // nx "sequence":
+		{"multiSig", false, false, "[case-2.0] Create MultiSig Tx for NFTs [Accept-token-ownership] - [internalTx3x_acceptTokenOwnership_err3] - wait-15-seconds.", "multisig-acc-nft-4", "100000000cin", 0, MultisigInfo{"create-internal-tx", "", "", 0, nil, "nft-grp-addr-4", 2, internalTx3x_acceptTokenOwnership_err3}, "MEMO : xxxxx", nil}, // nx "sequence":
+		{"multiSig", false, false, "[case-2.0] Sign MultiSig Tx for NFTs [Accept-token-ownership] - submit which signed by multisig-acc-nft-4 - wait-15-seconds.", "multisig-acc-nft-4", "100000000cin", 0, MultisigInfo{"multiSig-sign-tx", "", "", 0, nil, "nft-grp-addr-4", 0, internalTx3x_acceptTokenOwnership1}, "MEMO : Sign MultiSig Tx", nil},
```
Comment (b00f) [inline tests/multisigNFTs_test.go:177]: Need to wait for 15 seconds?

Comment (b00f) [reply tests/multisigNFTs_test.go:177]: If I am not mistaken only for approve process we need wait. Fo other transactions we don't have to wait.

File: tests/multisigNFTs_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+		{"multiSig", false, false, "[case-2.0] Sign MultiSig Tx for NFTs [Transfer-token-ownership] - submit which signed by multisig-acc-nft-3 - wait-15-seconds.", "multisig-acc-nft-3", "100000000cin", 0, MultisigInfo{"multiSig-sign-tx", "", "", 0, nil, "nft-grp-addr-2", 7, internalTx3x_transferTokenOwnership1}, "MEMO : Sign MultiSig Tx", nil},
+		{"nonFungibleToken", false, false, "[case-2.0] VERIFY nonfungible token transfer ownership [TNFT-PUBLIC-FALSE-02] - Happy path - wait-15-seconds.", "nft-mostafa", "0cin", 0, NonFungibleTokenInfo{"verify-transfer-token-ownership", "", "", "", "TNFT-PUBLIC-FALSE-02", "nft-mostafa", "nft-grp-addr-2", "token metadata", "", "", "", true, false, true, true, false, "nft-jeansoon", "0", "nft-carlo", "", "APPROVE_TRANFER_TOKEN_OWNERSHIP", "", "", []string{""}}, "[case-2.0] VERIFY nonfungible token transfer ownership [TNFT-PUBLIC-FALSE-02] - Happy path", nil},
```
Comment (b00f) [inline tests/multisigNFTs_test.go:170]: Only for approve_create_token and approve_transfer_ownership we have to wait.

**Comment by githubckgoh1439** [comment]
Hi, Mostafa

Will follow up those you commented above.

TQ.

File: tests/multisigNFTs_test.go (hunk 1 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go:84]: Hi, Mostafa

You are correct as this is redundant, as I planned to run this multisigNFTs_test.go as stand-alone process without involve other test-modules 1st, each time I testing new-cases.

Once fully completed this multisigNFTs_test.go later, that time only I merge with existing modules, which include the maintenance_test.go by remove the above case.

I take note your comment here.

TQ.

File: tests/multisigNFTs_test.go (hunk 2 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go:187]: Mostafa,

You are correct. Will remove the above once come to merging with other test-modules.

File: tests/multisigNFTs_test.go (hunk 3 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go:177]: Hi, Mostafa

Will test on this and revert.

File: tests/multisigNFTs_test.go (hunk 4 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go:170]: Will test on this and revert.

**Comment by githubckgoh1439** [comment]
Hi, Mostafa


This my testing-result & finding base on your comments :

Below with Test-cases that need the 5 seconds waiting-period
===============================================================


Pair-1 :
```
1. Action - Create MultiSig Tx for NFTs [Create-token]
2. Action - APPROVE NFTs token
-- If the above action Multisig-Create-Token not set-to-wait-5-seconds, this will have Error of 'Can not find this particular token-symbol'
```

Pair-2 :
```
3. Action - Create MultiSig Tx for NFTs [Mint-item]
4. Action - ENDORSE NFTs item 
-- If the above action Multisig-Mint-Item not set-to-wait-5-seconds, this will have Error of 'Code 2006 -Invalid token.'
```

Pair-3 :
```
5. Action - Create MultiSig Tx for NFTs [Transfer-token-ownership]
6. Action - VERIFY NFTs token transfer ownership 
-- If the above action Multisig-Transfer-token-ownership not set-to-wait-5-seconds, this will have Error of 'Code 2104 - Verify transfer token ownership invalid'
```

Remarks :
```
1. [APPROVE nonfungible token] : Is refer to your 'approve_create_token'
2. [VERIFY nonfungible token transfer ownership] : Is refer to your 'approve_transfer_ownership'
```


Please comments.

File: tests/multisigNFTs_test.go (hunk 4 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go]: Please see below latest update.

File: tests/multisigNFTs_test.go (hunk 3 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go]: Please see below latest update.

File: tests/multisigNFTs_test.go (hunk 2 above)
Comment (githubckgoh1439) [reply tests/multisigNFTs_test.go:187]: Mostafa,

I take note your comment here.

File: tests/tx_test.go [hunk 5]
```diff
@@ -39,14 +39,16 @@ func TestTxs(t *testing.T) {
...
 
-	tcs = append(tcs, makeBankTxs()...)
-	tcs = append(tcs, makeKycTxs()...)
-	tcs = append(tcs, makeMaintenaceTxs()...)
-	tcs = append(tcs, makeFeeTxs()...)
-	tcs = append(tcs, makeNonFungibleTokenTxs()...)
-	tcs = append(tcs, makeFungibleTokenTxs()...)
-	tcs = append(tcs, makeNameservicesTxs()...)
-	tcs = append(tcs, makeMultisigTxs()...)
+	// tcs = append(tcs, makeBankTxs()...)
```
Comment (b00f) [inline tests/tx_test.go:42]: Please make sure these lines are not commented in your last commit.

**Comment by githubckgoh1439** [comment]
> The case we discussed today is included?

Mostafa, 

I already updated test-cases base on last-week DEMO discussion under this commit:

https://github.com/maxonrow/maxonrow-go/pull/97/commits/ac7a4232aa519dfcdc0152476399ff53be30f24b

, which include:
case-2.1.1
case-2.1.2
case-2.1.3

File: tests/tx_test.go (hunk 5 above)
Comment (githubckgoh1439) [reply tests/tx_test.go]: Hi, Mostafa

I already done house-keeping base on this, which committed:

https://github.com/maxonrow/maxonrow-go/commit/ac7a4232aa519dfcdc0152476399ff53be30f24b


Please review.

**Comment by yenkhoon** [comment]
goh u might wanna enable back the test case before merging? anyway it looks good though thanks!

**Comment by githubckgoh1439** [comment]
Hi, Mostafa & YK

I will commit the latest test-cases for NFTs-multisig by today evening, before I come out the request for review and merging.

TQ.

**Pull request merged**